proposals/0000-runtime-environment-clear-prefix-and-storage-kill.md

```diff
@@ -0,0 +1,178 @@
+---
+Title: More fine grained control of `clear_prefix` and `storage_kill` from the runtime
+Number: 0
+Status: Proposed
+Authors:
+  - Bastian Köcher
+Created: 2023-02-27
+Category: Runtime Environment
+Requires:
+Replaces:
+---
+
+## Summary
+
+Give the runtime more fine grained control over the internals of `clear_prefix` and `storage_kill`. 
+We are proposing to introduce `ext_Storage_clear_prefix_version_3`, `ext_DefaultChildStorage_clear_prefix_version_3` 
+and `ext_DefaultChildStorage_storage_kill_version_3`. Notable changes are the introduction of a new 
+third parameter `maybe_cursor` and the change of the return value to `MultiRemovalResults`. 
+The `maybe_cursor` enables a runtime developer to start removing storage items from 
+a defined position. The new return value will give more details on how many items were 
+removed from the state and how many from the overlay etc.
+
+## Motivation
+
+Currently when a runtime developer calls `clear_prefix`/`kill_storage` with the
+same set of parameters in the same context it leads to the same keys being deleted.
+This can be quite confusing for runtime developers who are for example calling `clear_prefix`
+with some `limit` multiple times and expecting `limit * number of calls` keys to be deleted.
+While the actual result is that all calls remove exactly the same keys. They also don't get
+any information on how many keys got deleted in the state and how got deleted in the overlay.
+
+## Detailed Solution design
+
+Introduce the following new host functions:
+
+```rust
+fn ext_Storage_clear_prefix_version_3(prefix: u64, maybe_limit: u32, maybe_cursor: u64) -> u64;
+fn ext_DefaultChildStorage_clear_prefix_version_3(storage_key: u64, prefix: u64, maybe_limit: u32, maybe_cursor: u64) -> u64;
+fn ext_DefaultChildStorage_storage_kill_version_4(storage_key: u64, maybe_limit: u32, maybe_cursor: u64) -> u64;
```

Comment:
Is there a reason for the CamelCasing? All the other host functions are lower case

Reply:
Yeah, I trusted my own docs :P I will correct it, snake case is correct.